src/coreclr/src/vm/weakreferencenative.cpp

```diff
@@ -163,7 +164,7 @@ IWeakReference* GetComWeakReference(OBJECTREF* pObject)
         return nullptr;
     }
 
-    return pWeakReference.Extract();
+    return new NativeComWeakHandleInfo { pWeakReference.Extract(), wrapperId };
```

Comment:
This is a "throwing" `new`. Perhaps we should be using the loader allocator for this type of allocation? I am unclear what is the correct thing to do here. @jkotas Is that a valid concern or is this fine?

Reply:
What are the evaluation order guarantees for this in C++? The potential problem I see is that we extract the reference from `pWeakReference` before attempting to allocate `NativeComWeakHandleInfo`. If `NativeComWeakHandleInfo` allocation fails, the reference will leak.

Otherwise, this should work fine. We have "throwing" `new` over-writen to throw our own exceptions that get converted to managed exceptions on VM/managed boundaries.

Reply:
>What are the evaluation order guarantees for this in C++?

Boo. That is a good call. We need to defer that `Extract()` call until we know we have allocated the object.

>Otherwise, this should work fine. We have "throwing" new over-writen to throw our own exceptions that get converted to managed exceptions on VM/managed boundaries.

Excellent.